File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        mp = defaultdict(int)
        i, j , long_len , n = 0, 0, 0, len(nums)
        while j < n:
            mp[nums[j]] += 1
            while i < j and mp[nums[j]] > k:
                nums[i] -= 1
                i += 1
            long_len = max(long_len, j - i + 1)
            j += 1

        return long_len

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        int i = 0, j = 0, len = 0, n = nums.size();
        while (j < n) {
            mp[nums[j]]++;
            while (i < j && mp[nums[j]] > k) {
                mp[nums[i]]--;
                i++;
            }    
            len = max(len , j - i + 1);        
            j++;
        }
        return len;
    }
};